### ISSUE TITLE
How to pipe outputs of pose estimation into gvatrack

### ISSUE THREAD

**Issue by PulkitMishra** [dlstreamer/dlstreamer#315]
I wish to do pose estimation, convert the keypoints from pose estimation to get bounding boxes and then use these bounding boxes to do tracking (basically I wish to eliminate detection and replace it with pose estimation). Can someone point out as to how can this be done?

I am using this [sample](https://github.com/dlstreamer/dlstreamer/tree/master/samples/gstreamer/gst_launch/human_pose_estimation) as a starting point  and have been able to pipe outputs of pose estimation to gvapython module and converted the keypoints to bounding boxes. But from here, how do I pipe this to gvatrack?

**Comment by brmarkus** [comment]
Can you provide more details, maybe rephrase your idea?

**Comment by PulkitMishra** [comment]
Thanks for the response @brmarkus and apologies for not specifying clearly what I was looking for earlier.
Ideally to do object tracking we do detection first and then use outputs from detection to do tracking.
The issue is that I want to do pose estimation along with tracking. 

I could do pose estimation and then do object detection and use the outputs of object detection to do tracking. However, this would not be efficient as I would then be running two model inferences(pose estimation and object detection) which are not really needed as it is possible to simply get the output of object detection(i.e bounding boxes) from the output of pose estimation(i.e keypoints) itself. 

So what I would want to do is run pose estimation like [here](https://github.com/dlstreamer/dlstreamer/tree/master/samples/gstreamer/gst_launch/human_pose_estimation), convert these keypoints to bounding boxes by making use of gvapyhon callbacks and then use these bounding boxes to do tracking with gvatrack.


I have modified [pose estimation sample](https://github.com/dlstreamer/dlstreamer/tree/master/samples/gstreamer/gst_launch/human_pose_estimation) gotten all this to work but the tracking performance isn't great and I am unsure if I have done something wrong in my implementation or its an issue with gvatrack itself.



`#!/bin/bash
set -e

INPUT=${1:-https://github.com/intel-iot-devkit/sample-videos/raw/master/face-demographics-walking.mp4}

DEVICE=${2:-CPU}
TRACKING_TYPE=${4:-short-term-imageless} # Object tracking type, valid values: short-term-imageless, zero-term, zero-term-imageless

if [[ $3 == "display" ]] || [[ -z $3 ]]; then
  SINK_ELEMENT="gvawatermark ! videoconvert ! gvafpscounter ! autovideosink sync=false"
elif [[ $3 == "fps" ]]; then
  SINK_ELEMENT="gvafpscounter ! fakesink async=false "
else
  echo Error wrong value for SINK_ELEMENT parameter
  echo Possible values: display - render, fps - show FPS only
  exit
fi

if [[ $DEVICE == "CPU" ]]; then
  CONVERTER="videoconvert ! video/x-raw,format=BGRx"
  PREPROC_BACKEND="opencv"
elif [[ $DEVICE == "GPU" ]]; then
  CONVERTER="vaapipostproc ! video/x-raw\(memory:VASurface\)"
  PREPROC_BACKEND="vaapi-surface-sharing"
else
  echo Error wrong value for DEVICE parameter
  echo Possible values: CPU, GPU
  exit
fi

HPE_MODEL=human-pose-estimation-0001


if [[ $INPUT == "/dev/video"* ]]; then
  SOURCE_ELEMENT="v4l2src device=${INPUT}"
elif [[ $INPUT == *"://"* ]]; then
  SOURCE_ELEMENT="urisourcebin buffer-size=4096 uri=${INPUT}"
else
  SOURCE_ELEMENT="filesrc location=${INPUT}"
fi

PROC_PATH() {
    echo $(dirname "$0")/model_proc/$1.json
}

SCRIPTDIR="$(dirname "$(realpath "$0")")"
PYTHON_SCRIPT1=$SCRIPTDIR/postproc_callbacks/pose_to_bbox.py

HPE_MODEL_PATH=${MODELS_PATH}/intel/${HPE_MODEL}/FP16/${HPE_MODEL}.xml
HPE_MODEL_PROC=$(PROC_PATH $HPE_MODEL)
RECLASSIFY_INTERVAL=3

PIPELINE="gst-launch-1.0 $SOURCE_ELEMENT ! decodebin ! $CONVERTER ! \
gvaclassify model=$HPE_MODEL_PATH model-proc=$HPE_MODEL_PROC inference-interval=${RECLASSIFY_INTERVAL} device=$DEVICE inference-region=full-frame ! queue ! \
gvapython module=$PYTHON_SCRIPT1 ! queue ! \
gvatrack tracking-type=${TRACKING_TYPE} config=tracking_per_class=false ! queue ! \
$SINK_ELEMENT"

echo ${PIPELINE}
$PIPELINE`

`
import sys
import gi
gi.require_version('Gst', '1.0')
from gi.repository import Gst, GObject

from gstgva import VideoFrame

Gst.init(sys.argv)


def process_frame(frame: VideoFrame) -> bool:
    width = frame.video_info().width
    height = frame.video_info().height

    for tensor in frame.tensors():
        data = tensor.data()
        try:
            x_max = int((((data[5*2] if data[5*2] !=-1 else 0) + (data[6*2] if data[6*2] !=-1 else 0) + (data[7*2] if data[7*2] !=-1 else 0) + (data[11*2] if data[11*2] !=-1 else 0) + (data[12*2] if data[12*2] !=-1 else 0) + (data[13*2] if data[13*2] !=-1 else 0) + (data[15*2] if data[15*2] !=-1 else 0) + (data[17*2] if data[17*2] !=-1 else 0))/((1 if data[5*2] !=-1 else 0) + (1 if data[6*2] !=-1 else 0) + (1 if data[7*2] !=-1 else 0) + (1 if data[11*2] !=-1 else 0) + (1 if data[12*2] !=-1 else 0) + (1 if data[13*2] !=-1 else 0) + (1 if data[15*2] !=-1 else 0) + (1 if data[17*2] !=-1 else 0)))*width*1.15)
            y_min = int((((data[0*2 + 1] if data[0*2 + 1] !=-1 else 0) + (data[14*2 + 1] if data[14*2 + 1] !=-1 else 0) + (data[15*2 + 1] if data[15*2 + 1] !=-1 else 0) + (data[16*2 + 1] if data[16*2 + 1] !=-1 else 0) + (data[17*2 + 1] if data[17*2 + 1] !=-1 else 0))/((1 if data[0*2 + 1] !=-1 else 0) + (1 if data[14*2 + 1] !=-1 else 0) + (1 if data[15*2 + 1] !=-1 else 0) + (1 if data[16*2 + 1] !=-1 else 0) + (1 if data[17*2 + 1] !=-1 else 0)))*height*0.85)
            x_min = int((((data[2*2] if data[2*2] !=-1 else 0) + (data[3*2] if data[3*2] !=-1 else 0) + (data[4*2] if data[4*2] !=-1 else 0) + (data[8*2] if data[8*2] !=-1 else 0) + (data[9*2] if data[9*2] !=-1 else 0) + (data[10*2] if data[10*2] !=-1 else 0) + (data[14*2] if data[14*2] !=-1 else 0) + (data[16*2] if data[16*2] !=-1 else 0))/((1 if data[2*2] !=-1 else 0) + (1 if data[3*2] !=-1 else 0) + (1 if data[4*2] !=-1 else 0) + (1 if data[8*2] !=-1 else 0) + (1 if data[9*2] !=-1 else 0) + (1 if data[10*2] !=-1 else 0) + (1 if data[14*2] !=-1 else 0) + (1 if data[16*2] !=-1 else 0)))*width*0.85)
            y_max = int((((data[10*2 + 1] if data[10*2 + 1] !=-1 else 0) + (data[13*2 + 1] if data[13*2 + 1] !=-1 else 0))/((1 if data[10*2 + 1] !=-1 else 0) + (1 if data[13*2 + 1] !=-1 else 0)))*height*1.15)
            roi = frame.add_region(x_min, y_min, x_max - x_min, y_max - y_min)
        except:
            pass
    return True
    
 
`
